### PR TITLE
Add "Manager" Role to Delegate Contract

### PR DESCRIPTION
**Pull request by ejwessel**
## Description 
In reference to https://github.com/airswap/airswap-protocols/issues/154
If a Manager is to act on behalf of an owner in order to add a rule and set an intent through a Peer (delegate). Peer needs to have access control.

- the contract should have an owner
- the contract should have a list of parties that can set and unset an intent; a whitelist.
- owner is always apart of whitelist

## Coverage
Coverage Before
```
-----------------------|----------|----------|----------|----------|----------------|
File                   |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------|----------|----------|----------|----------|----------------|
 contracts/            |      100 |    95.83 |      100 |      100 |                |
  Imports.sol          |      100 |      100 |      100 |      100 |                |
  Peer.sol             |      100 |    95.83 |      100 |      100 |                |
 contracts/interfaces/ |      100 |      100 |      100 |      100 |                |
  IPeer.sol            |      100 |      100 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
All files              |      100 |    95.83 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
```
Coverage After
```
-----------------------|----------|----------|----------|----------|----------------|
File                   |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------|----------|----------|----------|----------|----------------|
 contracts/            |      100 |      100 |      100 |      100 |                |
  Imports.sol          |      100 |      100 |      100 |      100 |                |
  Peer.sol             |      100 |      100 |      100 |      100 |                |
 contracts/interfaces/ |      100 |      100 |      100 |      100 |                |
  IPeer.sol            |      100 |      100 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
All files              |      100 |      100 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
```

### CODE REVIEW

**Comment by ejwessel** [comment]
While this is related to https://github.com/airswap/airswap-protocols/pull/144
I did this as a separate PR since I felt it was an addition to the Peer and not necessarily tied to Manager.

**Comment by ejwessel** [comment]
I did not use openZeppelin due to not wanting to have it be possible to have multiple WhitelistAdmins

File: protocols/peer/contracts/Peer.sol [hunk 1]
```diff
@@ -50,10 +53,59 @@ contract Peer is IPeer, Ownable {
...
+    return whitelist[addressToCheck];
+  }
+
+  /**
+    * @notice adds to the list of whitelisted accounts that can interact with this peer
+    * @dev only callable by the owner of the contract
+    * @param addressToAdd the address to add to the whitelist
+    */
+  function addToWhitelist(address addressToAdd) external onlyOwner {
+    //TODO: emit
```
Comment (ejwessel) [inline protocols/peer/contracts/Peer.sol]: coming soon

File: protocols/peer/contracts/Peer.sol [hunk 2]
```diff
@@ -50,10 +53,59 @@ contract Peer is IPeer, Ownable {
...
+    whitelist[addressToAdd] = true;
+  }
+
+  /**
+    * @notice removes from the list of whitelisted accounts that can interact with this peer
+    * @dev only callable by the owner of the contract
+    * @param addressToRemove the address to add to the whitelist
+    */
+  function removeFromWhitelist(address addressToRemove) external onlyOwner {
+    //TODO: emit
```
Comment (ejwessel) [inline protocols/peer/contracts/Peer.sol]: coming soon

File: protocols/peer/contracts/Peer.sol [hunk 3]
```diff
@@ -50,10 +56,59 @@ contract Peer is IPeer, Ownable {
...
+  }
+
+  /**
+    * @notice removes from the list of whitelisted accounts that can interact with this peer
+    * @dev only callable by the owner of the contract
+    * @param addressToRemove the address to add to the whitelist
+    */
+  function removeFromWhitelist(address addressToRemove) external onlyOwner {
+    require(addressToRemove != owner(), "OWNER_MUST_BE_WHITELISTED");
+    whitelist[addressToRemove] = false;
```
Comment (poiresel) [inline protocols/peer/contracts/Peer.sol]: reason for choosing setting to false versus del whitelist[addressToRemove]

File: protocols/peer/contracts/Peer.sol [hunk 4]
```diff
@@ -50,10 +56,59 @@ contract Peer is IPeer, Ownable {
...
+  function transferOwnership(address newOwner) public onlyOwner {
+    whitelist[newOwner] = true;
+    whitelist[owner()] = false;
+    super.transferOwnership(newOwner);
+  }
+
+  /**
+    * @dev only whitelisted ensures that only whitelisted parties can call the method it modifies
+    */
+  modifier onlyWhitelisted() {
```
Comment (poiresel) [inline protocols/peer/contracts/Peer.sol]: put modifier at the top of the function
also pretty confident you can just call require(whitelist[msg.sender] , "CALLER..")

File: protocols/peer/contracts/Peer.sol [hunk 5]
```diff
@@ -50,10 +56,59 @@ contract Peer is IPeer, Ownable {
...
+    emit WhitelistRemoved(addressToRemove);
+  }
+
+  /**
+    * @notice transfers ownership to the new owner and ensures that whitelist is updated
+    * @dev only callable by the owner of the contract
+    * @param newOwner the address of the new owner of the contract
+    */
+  function transferOwnership(address newOwner) public onlyOwner {
+    whitelist[newOwner] = true;
```
Comment (poiresel) [inline protocols/peer/contracts/Peer.sol]: check that newOwner is not 0x0

File: protocols/peer/contracts/Peer.sol (hunk 4 above)
Comment (alicevhenshaw) [reply protocols/peer/contracts/Peer.sol]: you mean top of the contract @poiresel ?
agree with the removal of `== true`

File: protocols/peer/contracts/Peer.sol [hunk 6]
```diff
@@ -50,10 +56,59 @@ contract Peer is IPeer, Ownable {
   ) public {
     swapContract = ISwap(_swapContract);
     if (_peerContractOwner != address(0)) {
-      transferOwnership(_peerContractOwner);
+      whitelist[_peerContractOwner] = true;
```
Comment (alicevhenshaw) [inline protocols/peer/contracts/Peer.sol]: if _peerContractOwner == address(0) you still need to whitelist the owner. I would recommend removing this line, and instead adding the following _after_ the if statement (at the end of the constructor):
`whitelist[owner()] = true;`
This catches both cases.

File: protocols/peer/contracts/Peer.sol [hunk 7]
```diff
@@ -50,10 +56,59 @@ contract Peer is IPeer, Ownable {
...
+      super.transferOwnership(_peerContractOwner);
     }
   }
 
+  /**
+    * @notice determines if an address to interact with this peer
+    * @param addressToCheck the address to check if whitelisted or not 
+    */
+  function isWhitelisted(address addressToCheck) external returns(bool) {
```
Comment (alicevhenshaw) [inline protocols/peer/contracts/Peer.sol]: Insert a space after `returns`

Comment (alicevhenshaw) [reply protocols/peer/contracts/Peer.sol]: All of the new functions you've added do not have `_` on each parameter. This is inconsistent with the rest of the contract.

File: protocols/peer/contracts/Peer.sol [hunk 8]
```diff
@@ -30,12 +30,18 @@ import "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+  event WhitelistRemoved(address indexed account);
+
   // Swap contract to be used to settle trades
   ISwap public swapContract;
 
   // Mapping of takerToken to makerToken for rule lookup
   mapping (address => mapping (address => Rule)) public rules;
 
+  // Mapping of whitelisted addresses that can call on behalf of owner
+  mapping (address => bool) private whitelist;
```
Comment (ejwessel) [inline protocols/peer/contracts/Peer.sol]: https://github.com/airswap/airswap-protocols/issues/154#issuecomment-537483897

File: protocols/peer/contracts/Peer.sol (hunk 3 above)
Comment (ejwessel) [reply protocols/peer/contracts/Peer.sol]: no reason. good catch.

File: protocols/peer/contracts/Peer.sol [hunk 9]
```diff
@@ -39,11 +42,23 @@ contract Peer is IPeer, Ownable {
...
+  mapping (address => bool) private admins;
+
   // ERC-20 (fungible token) interface identifier (ERC-165)
   bytes4 constant internal ERC20_INTERFACE_ID = 0x277f8169;
 
+  /**
+    * @dev only admin ensures that only admin parties can call the method it modifies
+    */
+  modifier onlyAdmins() {
+    require(admins[msg.sender], "CALLER_NOT_ADMIN");
```
Comment (dmosites) [inline protocols/peer/contracts/Peer.sol]: `CALLER_MUST_BE_ADMIN`

File: protocols/peer/contracts/Peer.sol [hunk 10]
```diff
@@ -55,17 +70,61 @@ contract Peer is IPeer, Ownable {
...
+  function isAdmin(address _addressToCheck) view external returns (bool) {
+    return admins[_addressToCheck];
+  }
+
+  /**
+    * @notice adds to the list of admin accounts that can interact with this peer
+    * @dev only callable by the owner of the contract
+    * @param _addressToAdd the address to add to the admins
+    */
+  function addToAdmins(address _addressToAdd) external onlyOwner {
```
Comment (dmosites) [inline protocols/peer/contracts/Peer.sol]: Would suggest simply `addAdmin` and `removeAdmin`

File: protocols/peer/contracts/Peer.sol [hunk 11]
```diff
@@ -84,7 +143,7 @@ contract Peer is IPeer, Ownable {
     uint256 _maxTakerAmount,
     uint256 _priceCoef,
     uint256 _priceExp
-  ) external onlyOwner {
+  ) external onlyAdmins {
```
Comment (dmosites) [inline protocols/peer/contracts/Peer.sol]: The `only*` syntax is to enforce the _role_ of the caller, so feel `onlyAdmin` would be most correct.

**Pull request merged**